runners/direct-java/src/main/java/org/apache/beam/runners/direct/StructuralKey.java

```diff
@@ -26,52 +26,83 @@
  * A (Key, Coder) pair that uses the structural value of the key (as provided by
  * {@link Coder#structuralValue(Object)}) to perform equality and hashing.
  */
-class StructuralKey<K> {
+abstract class StructuralKey<K> {
+
+  private StructuralKey() {
+    // Do not extend outside of this class
+  }
+
+  public abstract K getKey();
+  /**
+   * Get the empty {@link StructuralKey}. All instances of the empty key are considered equal.
+   */
+  static StructuralKey<?> empty() {
+    return new StructuralKey<Object>() {
+      @Override
+      public Object getKey() {
+        return this;
+      }
+
+      @Override
+      public boolean equals(Object other) {
+        return other != null && getClass().isAssignableFrom(other.getClass());
+      }
+
+      @Override
+      public int hashCode() {
+        return getClass().hashCode();
+      }
+    };
+  }
+
   /**
    * Create a new Structural Key of the provided key that can be encoded by the provided coder.
    */
-  public static <K> StructuralKey<K> of(K key, Coder<K> coder) {
+  static <K> StructuralKey<K> of(K key, Coder<K> coder) {
     try {
-      return new StructuralKey<>(coder, key);
+      return new CoderStructuralKey<>(coder, key);
     } catch (Exception e) {
       throw new IllegalArgumentException(
           "Could not encode a key with its provided coder " + coder.getClass().getSimpleName(), e);
     }
   }
 
-  private final Coder<K> coder;
-  private final Object structuralValue;
-  private final byte[] encoded;
+  private static class CoderStructuralKey<K> extends StructuralKey<K> {
+    private final Coder<K> coder;
+    private final Object structuralValue;
+    private final byte[] encoded;
 
-  private StructuralKey(Coder<K> coder, K key) throws Exception {
-    this.coder = coder;
-    this.structuralValue = coder.structuralValue(key);
-    this.encoded = CoderUtils.encodeToByteArray(coder, key);
-  }
-
-  public K getKey() {
-    try {
-      return CoderUtils.decodeFromByteArray(coder, encoded);
-    } catch (CoderException e) {
-      throw new IllegalArgumentException(
-          "Could not decode Key with coder of type " + coder.getClass().getSimpleName());
+    private CoderStructuralKey(Coder<K> coder, K key) throws Exception {
+      this.coder = coder;
+      this.structuralValue = coder.structuralValue(key);
+      this.encoded = CoderUtils.encodeToByteArray(coder, key);
     }
-  }
 
-  @Override
-  public boolean equals(Object other) {
-    if (other == this) {
-      return true;
+    @Override
+    public K getKey() {
```

Comment:
It means that the runner only ever provides keys that have been deserialized from the initial representation it received. We only use it to deliver timers, but it means that timer deliveries are independent from the original key (as is appropriate).